zebra-rpc/src/queue.rs

```diff
@@ -0,0 +1,323 @@
+        // times, as the main loop also takes some time to run.
+        let extra_time = Duration::seconds(5);
+
+        let duration_to_expire =
+            Duration::seconds(NUMBER_OF_BLOCKS_TO_EXPIRE * spacing.num_seconds()) + extra_time;
+        let transactions = self.queue.transactions();
+        let now = Instant::now();
+
+        for tx in transactions.iter() {
+            let tx_time =
+                tx.1 .1
+                    .checked_add(
+                        duration_to_expire
+                            .to_std()
+                            .expect("should never be less than zero"),
+                    )
+                    .expect("this is low numbers, should always be inside bounds");
+
+            if now > tx_time {
+                self.queue.remove(*tx.0);
+            }
+        }
+    }
+
+    /// Remove transactions from the queue that had been inserted to the state or the mempool.
+    fn remove_committed(&mut self, to_remove: HashSet<UnminedTxId>) {
+        for r in to_remove {
+            self.queue.remove(r);
+        }
+    }
+
+    /// Check the mempool for given transactions.
+    ///
+    /// Returns transactions that are in the mempool.
+    async fn check_mempool<Mempool>(
+        mempool: Mempool,
+        transactions: HashSet<UnminedTxId>,
+    ) -> HashSet<UnminedTxId>
+    where
+        Mempool: Service<Request, Response = Response, Error = BoxError> + Clone + 'static,
+    {
+        let mut response = HashSet::new();
+
+        if !transactions.is_empty() {
+            let request = Request::TransactionsById(transactions);
+
+            let mempool_response = mempool
+                .oneshot(request)
+                .await
+                .expect("Requesting transactions should not panic");
+
+            match mempool_response {
+                Response::Transactions(txs) => {
+                    for tx in txs {
+                        response.insert(tx.id);
+                    }
+                }
+                _ => unreachable!("TransactionsById always respond with at least an empty vector"),
+            }
+        }
+
+        response
+    }
+
+    /// Check the state for given transactions.
+    ///
+    /// Returns transactions that are in the state.
+    async fn check_state<State>(
+        state: State,
+        transactions: HashSet<UnminedTxId>,
+    ) -> HashSet<UnminedTxId>
+    where
+        State: Service<ReadRequest, Response = ReadResponse, Error = zebra_state::BoxError>
+            + Clone
+            + Send
+            + Sync
+            + 'static,
+    {
+        let mut response = HashSet::new();
+
+        for t in transactions {
+            let request = ReadRequest::Transaction(t.mined_id());
+
+            let state_response = state
+                .clone()
+                .oneshot(request)
+                .await
+                .expect("Requesting transactions should not panic");
```

Comment:
Question: what if we change the state and it starts returning errors?

```suggestion
                .expect("Requesting transactions should not error");
```

Reply:
Let's ignore any errors returned by the state.

Reply:
Done as part of https://github.com/ZcashFoundation/zebra/pull/4015/commits/f82528104bd9b0c8573f196958f61cdd43a6652b